nemo/collections/asr/parts/k2/maploss.py

```diff
@@ -0,0 +1,269 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Copyright (c) 2020, Xiaomi CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, Union
+
+import k2
+import torch
+
+from nemo.collections.asr.parts.k2.autograd import sparse_abs
+from nemo.collections.asr.parts.k2.utils import (
+    create_sparse_wrapped,
+    create_supervision,
+    get_tot_objf_and_finite_mask,
+    load_graph,
+    make_blank_first,
+    prep_padded_densefsavec,
+    shift_labels_inpl,
+)
+from nemo.utils import logging
+
+
+class MAPLoss(torch.nn.Module):
+    """
+    Maximum a Posteriori Probability criterion.
+    It is implemented as Lattice-Free Maximum Mutual Information (LF-MMI) 
+    and LF-boosted-MMI (LF-bMMI) losses.
+    
+    Based on https://github.com/k2-fsa/snowfall/blob/master/snowfall/objectives/mmi.py
+    """
+
+    def __init__(
+        self,
+        num_classes: int,
+        blank: int,
+        reduction: str = "mean",
+        topo_type: str = "default",
+        topo_with_selfloops: bool = True,
+        loss_type: str = "mmi",
+        token_lm: Optional[Union[k2.Fsa, str]] = None,
+        intersect_pruned: bool = False,
+        boost_coeff: float = 0.0,
+        **kwargs,
+    ):
+        super().__init__()
+        self.num_classes = num_classes
```

Comment:
Check K2 import guard here.

Reply:
Done in [8d3e482](https://github.com/NVIDIA/NeMo/pull/3351/commits/8d3e48215629824b59a020526a0444c944b2d1c2)